src/PSR7/Validators/BodyValidator/MultipartValidation.php

```diff
@@ -0,0 +1,187 @@
+                        $validator->validate($headerValue, $headerSchema);
+                    } catch (SchemaMismatch $e) {
+                        throw InvalidHeaders::becauseValueDoesNotMatchSchemaMultipart($partName, $headerName, $headerValue, $addr, $e);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Prepare a Multipart message body (a set of normal parts) for validation against a schema
+     *
+     * @see https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html
+     * @see https://swagger.io/docs/specification/describing-request-body/multipart-requests/
+     *
+     * @return mixed[]
+     *
+     * @throws InvalidBody
+     * @throws TypeMismatch
+     */
+    private function parseMultipartData(OperationAddress $addr, StreamedPart $document) : array
+    {
+        $multipartData = []; // a buffer to fill up with message parts
+        foreach ($document->getParts() as $i => $part) {
+            $partContentType = $part->getHeader('Content-Type');
+
+            if (preg_match('#^application/.*json$#', $partContentType)) {
+                $partBody = json_decode($part->getBody(), true);
+                if (json_last_error() !== JSON_ERROR_NONE) {
+                    throw InvalidBody::becauseBodyIsNotValidJson(json_last_error_msg(), $addr);
+                }
+            } else {
+                $partBody = $part->getBody();
+            }
+
+            // if name is not set, it should be validated with "additionalProperties" keyword
+            $multipartData[$part->getName() ?? '____' . $i] = $partBody;
+        }
+
+        return $multipartData;
+    }
+
+    private function detectEncondingContentType(Encoding $encoding, StreamedPart $part, Schema $partSchema) : ?string
+    {
+        $contentType = $encoding->contentType;
+
+        if (! $contentType) {
+            // fallback strategy:
+            // @see https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#encoding-object
+            // @see https://swagger.io/docs/specification/describing-request-body/multipart-requests/
+            //
+            // Default value depends on the property type: for string with format being binary – application/octet-stream;
+            // for other primitive types – text/plain; for object - application/json; for array – the default is defined based on the inner type.
+            // The value can be a specific media type (e.g. application/json), a wildcard media type (e.g. image/*),
+            // or a comma-separated list of the two types.
+            if ($partSchema->type === 'string') {
+                if (in_array($partSchema->format, ['binary', 'base64'])) {
+                    $contentType = 'application/octet-stream';
+                } else {
+                    $contentType = 'text/plain';
+                }
+            } elseif (in_array($partSchema->type, ['object', 'array'])) {
+                $contentType = 'application/json';
```

Comment:
ehm. `xml`, `yaml` and others ?

Reply:
Very good point, it means the library needs the support of all possible formats out there? The usage of XML must be huge on the Internet these days. Just thinking about how to design this.

Reply:
In general this means we are unable to guess the content-type here. And should match more types if they are not specified in schmea

https://swagger.io/docs/specification/data-models/representing-xml/

So in general this could be any object-serializible format

If I understand correctly this example - you can use the same schema to xml and json depending on what content-type you have